File: authortex/authortex.pl
#!/usr/bin/perl

#Usage: ./authortex.pl authorlist [startindex 1] [sdssauthors] [sdssinstitutions]

#Output: {authorlist}.tex

#authorlist is a file containing the desired author names in the
#following format
#
#
#FIRST TIER
#Richards
#Fan
#VandenBerk
#Schneider
#Strauss
#
#SECOND TIER
#Gunn
#Pier
#Frieman
#Uomoto
#BahcallN
#BahcallJ
#
#
#The first tier authors need to be in the correct order.  The second
#tier authors are sorted alphabetically.
#
#[startindex] is an optional argument that allows the author footnotes
#to start at a number other than one.  For example if there is a
#footnote to the title.
#
#[sdssauthors] is the database file with author names and institutions
#
#[sdssinstitutions] is the database file with addresses of the institutions
#
#Note that these paths are hardcoded now!

my $def_authorfile = "./sdssauthors";
my $def_instfile   = "./sdssinstitutions";

if (defined $ARGV[2]) {
    $authorfile = $ARGV[2];
} else {
    $authorfile = $def_authorfile;
}
if (defined $ARGV[3]) {
    $institutionfile = $ARGV[3];
} else {
    $institutionfile = $def_instfile;
}

#fill author hash array
open(AUTHORFP,"<$authorfile") || die "cant open sdssauthors database file";
while($inline = <AUTHORFP>) {

    chop($inline);
    @authorline=split("/",$inline);
    $name = @authorline[0];
    
    $authors{$name} = [ @authorline ];

}
close(AUTHORFP);


#fill institution hash array
open(INSTFP,"<$institutionfile") || die "cant open sdssinstitutions database file";
while($inline = <INSTFP>) {

    chop($inline);
    @instline=split("/",$inline);
    $inst = @instline[0];
    
    $institutions{$inst} = [ @instline ];
    $institutions{$inst}[2] = 0;

}
close(INSTFP);

#Read authorlist input file
#output authorlist.tex

$tier1authorstring = "";
$tier1institutionstring = "";

$tier = 0;
if (defined $ARGV[1]) {
    $counter = $ARGV[1];
} else {
    $counter = 1;
}
open(INFP,$ARGV[0]) || die "cant open infile";
while($inline = <INFP>) {
    @input=split(" ",$inline);
    if ($inline =~ /^\s/) {
	next;
    }
    if (@input[0] eq "FIRST") {
	$tier = 1;
	next;
    } elsif (@input[0] eq  "SECOND") {
	$tier = 2;
	next;
    } else {
	$name = @input[0];
    }
    
    push(@allauthors,$name);

    if ($tier == 1) {
	if ($authors{$name}[1] eq "") {
	    die "No entry in author database for $name";
	}
	#$authorstring = "$authors{$name}[1]\\altaffilmark{";
	$authorstring = "$authors{$name}[1],\\altaffilmark{";
	$tier1authorstring = $tier1authorstring.$authorstring;

	@instlist = split(" ",$authors{$name}[2]);

	#Assign counter number to each institution
	foreach $inst (@instlist) {
	    if ($institutions{$inst}[1] eq "") {
		die "No entry in institution database for $inst";
	    }
	    if ($institutions{$inst}[2] == 0) {
		$institutions{$inst}[2] = $counter;
		$institutionstring = "\\altaffiltext{$institutions{$inst}[2]}{$institutions{$inst}[1]}\n";
		$tier1institutionstring = $tier1institutionstring.$institutionstring;
		$counter++;
	    }
	    push(@instcounterlist,$institutions{$inst}[2]);
	}
	if ($#instlist == 0) {
	    $institutionstring = pop(@instcounterlist);
	    $tier1authorstring = $tier1authorstring.$institutionstring;
	} else {
	    @instcounterlistsort = sort { $b <=> $a } @instcounterlist;
	    for $i (0 .. $#instcounterlistsort) {
		if ($i == 0) {
		    $institutionstring = pop(@instcounterlistsort);
		    $tier1authorstring = $tier1authorstring.$institutionstring;
		    pop(@instcounterlist);
		} else {
		    $institutionstring = pop(@instcounterlistsort);
		    $tier1authorstring = $tier1authorstring.",".$institutionstring;
		    pop(@instcounterlist);
		}
	    }
	}
	#$tier1authorstring = $tier1authorstring."},\n";
	$tier1authorstring = $tier1authorstring."}\n";
    }

    if ($tier == 2) {
	@input=split(" ",$inline);
	push(@tier2list,@input[0]);
    }

}
close(INFP);

$n = 0;
#Now take care of second tier authors, who need to be sorted
foreach $name (sort @tier2list) {

    if ($authors{$name}[1] eq "") {
        die "No entry in author database for $name";
    }
    #$authorstring = "$authors{$name}[1]\\altaffilmark{";
    if ($n != $#tier2list) {
        $authorstring = "$authors{$name}[1],\\altaffilmark{";
    } else {
        $authorstring = "and "."$authors{$name}[1]\\altaffilmark{";
    }
    $tier2authorstring = $tier2authorstring.$authorstring;

    @instlist = split(" ",$authors{$name}[2]);

    #Assign counter number to each institution
    foreach $inst (@instlist) {
        if ($institutions{$inst}[1] eq "") {
            die "No entry in institution database for $inst";
        }
        if ($institutions{$inst}[2] == 0) {
            $institutions{$inst}[2] = $counter;
            $institutionstring = "\\altaffiltext{$institutions{$inst}[2]}{$institutions{$inst}[1]}\n";
            $tier2institutionstring = $tier2institutionstring.$institutionstring;
            $counter++;
        }
        push(@instcounterlist,$institutions{$inst}[2]);
    }
    if ($#instlist == 0) {
        $institutionstring = pop(@instcounterlist);
        $tier2authorstring = $tier2authorstring.$institutionstring;
    } else {
        @instcounterlistsort = sort { $b <=> $a } @instcounterlist;
        for $i (0 .. $#instcounterlistsort) {
            if ($i == 0) {
                $institutionstring = pop(@instcounterlistsort);
                $tier2authorstring = $tier2authorstring.$institutionstring;
                pop(@instcounterlist);
            } else {
                $institutionstring = pop(@instcounterlistsort);
                $tier2authorstring = $tier2authorstring.",".$institutionstring;
                pop(@instcounterlist);
            }
        }
    }

    #Deal with last entry.  
    #Right now these are the same because of "and the SDSS Collaboration"
    if ($n != $#tier2list) {
        $tier2authorstring = $tier2authorstring."}\n";
    } else {
        $tier2authorstring = $tier2authorstring."}\n";
    }
    $n++;
}

$outfile = $ARGV[0].".tex";
open(OUTFP,">$outfile") || die "cant open output file";

printf(OUTFP "\\author{\n");
printf(OUTFP "$tier1authorstring");
printf(OUTFP "$tier2authorstring");
printf(OUTFP "}\n");
#printf(OUTFP "and the SDSS Collaboration}\n");

printf(OUTFP "\n");
printf(OUTFP "$tier1institutionstring");
printf(OUTFP "$tier2institutionstring");

close(OUTFP);

#foreach $author (sort @allauthors) {
#    print "$author,";
#}
#print "\n";

File: authortex/authortex.py
#!/usr/bin/env python
"""
Usage: 
    authortex.py [-a|--authors authordb] [-i|--institutions institutiondb] authorlist


Options:
  -h, --help            show this help message and exit
  -a AUTHORS, --authors=AUTHORS
                        Author Database. Default ./authors
  -i INST, --institutions=INST
                        Institution Database. Default ./institutions

    Two tiers of authors, a first tier ordered as given, and a second ordered
    by the identifier given (usually a last name).

This program facilitates the use of author and institution databases. The
author list can be split into two tiers.  The first tier is ordered as
presented, the second tier is ordered alphabetically by the identifier. The
author list format and the author/institution database formats are taken from
Michael Blanton's, as used by his authortex perl code.

The author and inst. database files can be indicated in the command line
options, but by default the files authors and institutions are looked for and
used.

The results are sent to standard output.


Example author list:
-------------------------------

FIRST TIER
Karabekian
Yossarian
Gandalf

SECOND TIER
SmithJ
SmithR
Jones


Example author database:  AuthorKey/Full Name/Institution List
Each author can have multiple institutions
--------------------------------
Karabekian/Rabo Karabekian/Midland
Yossarian/John Yossarian/Pianosa
....


Example institution database:  InstitutionKey/Full Affiliation Info
------------------------------------
Midland/Midland Institute, Midland City, Some City, Some State, 12345
Pianosa/Pianosa University, Pianosa, Italy
....

Author and institution keys are case insensitive, as are the "first tier" and
"second tier" tokens.




Modification History
--------------------------

2007-09-04  Created. Designed to work with Michael Blanton's author and
            institution database formats. Erin Sheldon, NYU

"""

import sys
from optparse import OptionParser

# Default file names
def_authorfname = "./authors"
def_instfname = "./institutions"

# Options and usage
usage = """
    authortex.py [-a|--authors authordb] [-i|--institutions institutiondb] authorlist
"""
parser=OptionParser(usage)
parser.add_option("-a","--authors", 
                  dest="authors",
                  help="Author Database. Default ./authors", 
                  default=def_authorfname)
parser.add_option("-i","--institutions", 
                  dest="inst",
                  help="Institution Database. Default ./institutions", 
                  default=def_instfname)

def ParseAuthorDb(authorfile):
    """ Parse the author database"""

    authors = {}
    for line in authorfile:

        sp = line.split('/')
        if len(sp) == 3:

            # There can be multiple institutions, so use split to create
            # a list
            author_id = sp[0].strip().lower()
            name = sp[1].strip()
            inst_idlist = sp[2].strip().split()
            inst_idlist = [id.lower() for id in inst_idlist]

            authors[author_id] = \
                    {'author_id':author_id, 
                     'name':name, 
                     'inst_idlist':inst_idlist}

    return authors

def ParseInstitutionDb(instfile):
    """Parse the institution database"""
    inst = {}
    for line in instfile:

        # Allow / in name. Just split on first occurence
        id = line.find('/')
        if id != -1:
            inst_id = line[0:id].strip().lower()
            affil = line[id+1:].strip()

            inst[inst_id] = {'inst_id':inst_id, 'affil':affil}
    return inst

def ParseAuthorlist(authorfile):
    """Parse the user's author list"""
    
    first = []
    second = []
    isfirst=0
    issec=0
    for line in authorfile:
        line = line.strip()
        sp = line.split()
        if len(sp) == 2:
            # First or Second indicator is given
            if sp[0].upper() == 'FIRST':
                isfirst=1
                issec=0
            elif sp[0].upper() == 'SECOND':
                issec=1
                isfirst=0
        elif len(sp) == 1:
            id = sp[0].lower()
            if isfirst:
                first.append(id)
            elif issec:
                second.append(id)

    # sort by the identifier
    second.sort()

    # Now append the seconds to the first list
    for s in second:
        first.append(s)
    return first

def ProcessAuthors(all_authors, all_inst, authorlist):
    """
    Loop through the requested author ids and keep track of the
    affiliations in the order they come
    """

    kept_authors = []
    author_marks = []
    kept_inst = []
    kept_inst_marks = {}
    marknum = 1
    for author_id in authorlist:

        if author_id not in all_authors:
            print 'Author id: "'+author_id+'" not in database'
            sys.exit(45)
        else:
            name = all_authors[author_id]['name']
            inst_idlist = all_authors[author_id]['inst_idlist']

        # Keep a unique, ordered list of institutions
        amarks=[]
        for inst_id in inst_idlist:
            if inst_id not in all_inst:
                print 'Institution id: "'+inst_id+'" not in database'
                sys.exit(45)
            else:
                output = name
                if inst_id not in kept_inst_marks:
                    affil = all_inst[inst_id]['affil']
                    mark=str(marknum)
                    kept_inst.append({'mark':mark, 'affil':affil})

                    kept_inst_marks[inst_id] = mark

                    marknum=marknum+1


                # get the mark for this id
                amark = kept_inst_marks[inst_id]
                amarks.append(amark)

        if len(amarks) > 0:
            # Sort numerically (key=int uses int(mark) as sort key)
            amarks.sort(key=int)
            # Note comma
            afm = ',\\altaffilmark{'+','.join(amarks)+'}'
            author_marks.append(afm)
            kept_authors.append(name)

    # Combine, but with the marks *after* the comma.  This is less clean
    # to code, but looks better
    if len(kept_authors) > 0:
        # fix comma on last name
        author_marks[-1] = author_marks[-1].replace(',','',1)
        # Combine
        kept_authors = [k+a for k,a in zip(kept_authors,author_marks)]
        
    
    return kept_authors, kept_inst


def PrintAuthors(kept_authors):
    """Print the authorlist"""

    if len(kept_authors) == 0:
        print 'No good authors/institutions found'
        sys.exit(45)

    # Create the text for the authors
    if len(kept_authors) > 1:
        kept_authors[-1] = 'and '+kept_authors[-1]

    kept_authors = '\n'.join(kept_authors)

    # fix last comma


    # Print out the authorlist with affilmarks
    print '\\author{'
    print kept_authors
    print '}'
    print

def PrintAffil(kept_inst):
    """Print the affiliatons"""
    # Print out the affiltext commands
    affiltext = ''
    for inst in kept_inst:
        affiltext = '\\altaffiltext'
        affiltext = affiltext + '{'+inst['mark']+'}'
        affiltext = affiltext + '{'+inst['affil']+'}'

        print affiltext

def GetFiles(argv):
    """
    Extract the filenames from the command line, open the files and return
    the file objects
    """
    (options, args) = parser.parse_args(argv[1:])
    if len(args) < 1:
        parser.print_help()
        sys.exit(45)

    # Optional arguments
    authorfname = options.authors
    instfname = options.inst
    # required argument: the user's author list
    myauthorfname = args[0]

    # open the files
    authorfile = open(authorfname, "r")
    instfile = open(instfname,"r")
    myauthorfile = open(myauthorfname)
    
    return authorfile, instfile, myauthorfile


#
# Main program
#

authorfile, instfile, myauthorfile = GetFiles(sys.argv)

# Parse the databases
all_authors = ParseAuthorDb(authorfile)
all_inst = ParseInstitutionDb(instfile)

# Parse the users authorlist
authorlist = ParseAuthorlist(myauthorfile)

# Process the authors, checking for matches in the
# author and institution databases
kept_authors, kept_inst = ProcessAuthors(all_authors, 
                                         all_inst,
                                         authorlist)
# To stdout
PrintAuthors(kept_authors)
PrintAffil(kept_inst)


